File: PropertiesDrivenDevelopment/elixir/checkout/lib/checkout.ex
#---
# Excerpted from "Property-Based Testing with PropEr, Erlang, and Elixir",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material,
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose.
# Visit http://www.pragmaticprogrammer.com/titles/fhproper for more book information.
#---
defmodule Checkout do
  @moduledoc """
  A Checkout counter implementation
  """

  def valid_price_list(list) do
    sorted = Enum.sort(list)
    length(list) == length(Enum.dedup_by(sorted, fn {x, _} -> x end))
  end

  def valid_special_list(list) do
    sorted = Enum.sort(list)

    Enum.all?(list, fn {_, x, _} -> x != 0 end) &&
      length(list) == length(Enum.dedup_by(sorted, fn {x, _, _} -> x end))
  end


  def total(item_list, price_list, specials) do
    if not valid_price_list(price_list) do #(1)
      raise RuntimeError, message: "invalid list of prices"
    end

    if not valid_special_list(specials) do
      raise RuntimeError, message: "invalid list of specials"
    end

    counts = count_seen(item_list)
    {counts_left, prices} = apply_specials(counts, specials)
    prices + apply_regular(counts_left, price_list)
  end


  defp count_seen(item_list) do
    count = fn x -> x + 1 end

    Map.to_list(
      Enum.reduce(item_list, Map.new(), fn item, m ->
        Map.update(m, item, 1, count)
      end)
    )
  end

  defp apply_specials(items, specials) do
    Enum.map_reduce(items, 0, fn {name, count}, price ->
      case List.keyfind(specials, name, 0) do
        nil ->
          {{name, count}, price}

        {_, needed, value} ->
          {{name, rem(count, needed)}, value * div(count, needed) + price}
      end
    end)
  end

  defp apply_regular(items, price_list) do
    Enum.sum(
      for {name, count} <- items do
        count * cost_of_item(price_list, name)
      end
    )
  end

  defp cost_of_item(price_list, name) do
    case List.keyfind(price_list, name, 0) do
      nil -> raise RuntimeError, message: "unknown item: #{name}"
      {_, price} -> price
    end
  end

  ## Old work-in-progress implementations
  defmodule V1 do
    def total(item_list, price_list, _specials) do
      Enum.sum(
        for item <- item_list do
          elem(List.keyfind(price_list, item, 0), 1)
        end
      )
    end

  end

  defmodule V2 do
    def total(item_list, price_list, specials) do
      counts = count_seen(item_list)
      {counts_left, prices} = apply_specials(counts, specials)
      prices + apply_regular(counts_left, price_list)
    end


    defp count_seen(item_list) do
      count = fn x -> x + 1 end

      Map.to_list(
        Enum.reduce(item_list, Map.new(), fn item, m ->
          Map.update(m, item, 1, count)
        end)
      )
    end


    defp apply_specials(items, specials) do
      Enum.map_reduce(items, 0, fn {name, count}, price ->
        case List.keyfind(specials, name, 0) do
          nil ->
            {{name, count}, price} #(2)

          {_, needed, value} ->
            {{name, rem(count, needed)}, #(3)
              value * div(count, needed) + price} #(4)
        end
      end)
    end

    defp apply_regular(items, price_list) do
      Enum.sum(
        for {name, count} <- items do
          {_, price} = List.keyfind(price_list, name, 0)
          count * price
        end
      )
    end

  end

  defmodule V3 do
    def valid_special_list(list) do
      Enum.all?(list, fn {_, x, _} -> x != 0 end)
    end

    def total(item_list, price_list, specials) do
      if not valid_special_list(specials) do #(5)
        raise RuntimeError, message: "invalid list of specials"
      end

      counts = count_seen(item_list)
      {counts_left, prices} = apply_specials(counts, specials)
      prices + apply_regular(counts_left, price_list)
    end


    defp count_seen(item_list) do
      count = fn x -> x + 1 end

      Map.to_list(
        Enum.reduce(item_list, Map.new(), fn item, m ->
          Map.update(m, item, 1, count)
        end)
      )
    end


    defp apply_specials(items, specials) do
      Enum.map_reduce(items, 0, fn {name, count}, price ->
        case List.keyfind(specials, name, 0) do
          nil ->
            {{name, count}, price}

          {_, needed, value} ->
            {{name, rem(count, needed)}, value * div(count, needed) + price}
        end
      end)
    end


    defp apply_regular(items, price_list) do
      Enum.sum(
        for {name, count} <- items do
          count * cost_of_item(price_list, name)
        end
      )
    end

    defp cost_of_item(price_list, name) do
      case List.keyfind(price_list, name, 0) do
        nil -> raise RuntimeError, message: "unknown item: #{name}"
        {_, price} -> price
      end
    end

  end
end

File: PropertiesDrivenDevelopment/erlang/checkout/src/checkout.erl
%% ---
%%  Excerpted from "Property-Based Testing with PropEr, Erlang, and Elixir",
%%  published by The Pragmatic Bookshelf.
%%  Copyrights apply to this code. It may not be used to create training material,
%%  courses, books, articles, and the like. Contact us if you are in doubt.
%%  We make no guarantees that this code is fit for any purpose.
%%  Visit http://www.pragmaticprogrammer.com/titles/fhproper for more book information.
%%---
-module(checkout).
-type item() :: string().
-type price() :: integer().
-type special() :: {item(), pos_integer(), price()}.

-export([valid_price_list/1, valid_special_list/1, total/3]).

-spec valid_price_list([{item(), price()}]) -> boolean().
valid_price_list(List) ->
    length(List) =:= length(lists:ukeysort(1, List)).

-spec valid_special_list([special()]) -> boolean().
valid_special_list(List) ->
    lists:all(fun({_,X,_}) -> X =/= 0 end, List) andalso
    length(List) =:= length(lists:ukeysort(1, List)).

-spec total([item()], [{item(), price()}], [special()]) -> price().
total(ItemList, PriceList, Specials) ->
    valid_price_list(PriceList) orelse error(invalid_price_list), %(1)
    valid_special_list(Specials) orelse error(invalid_special_list),
    Counts = count_seen(ItemList),
    {CountsLeft, Prices} = apply_specials(Counts, Specials),
    Prices + apply_regular(CountsLeft, PriceList).

-spec count_seen([item()]) -> [{item(), pos_integer()}].
count_seen(ItemList) ->
    Count = fun(X) -> X+1 end,
    maps:to_list(
      lists:foldl(fun(Item, M) -> maps:update_with(Item, Count, 1, M) end,
                  maps:new(), ItemList)
    ).

-spec apply_specials([{item(), pos_integer()}], [special()]) ->
    {[{item(), pos_integer()}], price()}.
apply_specials(Items, Specials) ->
    lists:mapfoldl(fun({Name, Count}, Price) ->
        case lists:keyfind(Name, 1, Specials) of
            false -> % not found
                {{Name, Count}, Price};
            {_, Needed, Value} ->
                {{Name, Count rem Needed},
                 Value * (Count div Needed) + Price}
        end
    end, 0, Items).

-spec apply_regular([{item(), integer()}], [{item(), price()}]) -> price().
apply_regular(Items, PriceList) ->
    lists:sum([Count * cost_of_item(Name, PriceList)
               || {Name, Count} <- Items]).

cost_of_item(Name, PriceList) ->
    case proplists:get_value(Name, PriceList) of
        undefined -> error({unknown_item, Name});
        Price -> Price
    end.

